src/main/java/org/zalando/nakadi/service/EventStreamWriterBinary.java

```diff
@@ -44,11 +63,59 @@ public int writeBatch(final OutputStream os, final Cursor cursor, final List<Str
         final byte[] offset = cursor.getOffset().getBytes(StandardCharsets.UTF_8);
         os.write(offset);
         byteCount += offset.length;
+
+        os.write(B_CURSOR_PARTITION_END);
+        if (!events.isEmpty()) {
+            os.write(B_EVENTS_ARRAY_BEGIN);
+            for (int i = 0; i < events.size(); i++) {
+                final byte[] event = events.get(i);
+                os.write(event);
+                byteCount += event.length;
+                if (i < (events.size() - 1)) {
+                    os.write(B_COMMA_DELIM);
+                } else {
+                    os.write(B_CLOSE_BRACKET);
+                }
+            }
+        }
```

Comment:
In my first proposal I avoided duplication but had to do some type casting. After some feedback, I decided that duplicating would be better than casting. @antban you can take a look at the `extractor` thing https://github.com/zalando/nakadi/pull/673/commits/687dc11690fdc5cb4eddf6d7b0193be6d1d93741

Another approach was proposed by @adyach but it would require creating a lambda with a close on every call to this method. Since it's a hotspot and this change is about optimization, I thought it would be acceptable to have some duplication.